post-build-hook-queue/src/main.rs

```diff
@@ -57,57 +59,103 @@ fn main() -> anyhow::Result<()> {
         .try_clone_to_owned()
         .context("Failed to convert stdin to an owned fd")?;
 
-    let sock: UnixDatagram = UnixDatagram::from(stdin_fd);
+    let workers: usize = std::env::var("NBPHQ_WORKERS")
+        .as_deref()
+        .unwrap_or("4")
+        .parse()?;
+    let queue_size: usize = std::env::var("NBPHQ_QUEUE_SIZE")
+        .as_deref()
+        .unwrap_or("64")
+        .parse()?;
+
+    let (sender, receiver) = std::sync::mpsc::sync_channel::<OsString>(queue_size);
+    let receiver = Arc::new(Mutex::new(receiver));
 
+    let mut worker_threads = Vec::<JoinHandle<()>>::new();
+    for _ in 0..workers {
+        let receiver = receiver.clone();
+        worker_threads.push(std::thread::spawn(move || {
+            while let Ok(path) = {
+                let receiver = receiver.lock().unwrap();
+                receiver.recv()
+            } {
+                if let Err(e) = try_push_path(&path) {
+                    log::error!(
+                        "Push path failed for path {}: {}",
+                        String::from_utf8_lossy(path.as_bytes()),
+                        e
+                    );
+                }
+            }
+        }));
+    }
+
+    let sock: UnixDatagram = UnixDatagram::from(stdin_fd);
     // A store path over 4096 characters would be insane, right?
     let mut buf: Vec<u8> = vec![0; 4096];
     loop {
         let n_bytes: usize = match sock.recv(&mut buf) {
             Ok(n_bytes) => n_bytes,
-            Err(e) => match e.kind() {
-                ErrorKind::WouldBlock => {
-                    // service will be started by systemd when there is more data
-                    std::process::exit(0);
+            Err(e) => {
+                let code = match e.kind() {
+                    ErrorKind::WouldBlock => {
+                        // service will be started by systemd when there is more data
+                        0
+                    }
+                    _ => {
+                        log::error!("Failed to recv from socket: {e:?}");
+                        1
+                    }
+                };
+                drop(sock);
+                drop(sender);
```

Comment:
Won't the child workers panic when the sender is dropped?

https://doc.rust-lang.org/std/sync/mpsc/index.html#disconnection